### PR TITLE
Refactor velocity representations as integers

### PR DESCRIPTION
**Pull request by flferretti**
This pull request refactors the velocity representation in the JaxSim API. The changes include avoiding the use of `enum` in `VelRepr` and making `velocity_representation` a non-static argument in some methods of `jaxsim.api.JaxSimModelData`. These changes improve the compatibility with JAX while avoiding breaking changes in the API, making it possible to potentially using `jax.vmap` on different velocity representations

----
📚 Documentation preview 📚: https://jaxsim--160.org.readthedocs.build//160/

### CODE REVIEW

**Comment by flferretti** [comment]
The AD tests take a bit longer to complete. This can be due to the JAX traceback used in `JaxSimModelReferences` that can lead to this effect when trying to compute the gradients

**Comment by diegoferigo** [comment]
Awesome, thanks @flferretti for this PR. I'd expect the following consequences of this PR:

- With static velocity representations, functions compiled passing different values are for JAX different. Only the section of the removed match-case statement is included in the compiled IR.
- I expect now a slightly slower compilation of our functions since they include more code, but in most cases the match-case statements only contain a bunch of transforms computation, nothing too heavy. `TL;DR` slower JIT compilations for a single call, but no recompilations in case multiple representations are needed.
- I expect, however, a smaller "equivalent binary" size of a large jit-compiled application. In fact, it happens to call the same JIT-compiled function with different representations (sometimes making calculation e.g. in body-fixed is easier even if the active representation is different), therefore before the full binary might have contained multiple copies of the same RBDA where the only difference was the match-case entry used to adjust the representation. `TL;DR` better memory footprint.

Not yet sure if my intuition is correct.

This being said, I'd like to tag a release with all the previous improvement. I don't expect surprises here, but being such a large change touching pretty much all our API surface, I prefer being cautious and include this PR in the following release (v0.4.0). I'll be reviewing this shortly.

**Comment by diegoferigo** [comment]
> The AD tests take a bit longer to complete.

I guess that now AD needs to propagate gradients through all possible branches instead of just one. And yes, this might take longer.

**Comment by flferretti** [comment]
I agree with your intuitions, the IR should now include the three branches for each velocity representation and the recompilation should not be triggered when we use `switch_velocity_representation` or `from_other_to_inertial` or `from_inertial_to_other`.

>  I'd like to tag a release with all the previous improvement. I don't expect surprises here, but being such a large change touching pretty much all our API surface, I prefer being cautious and include this PR in the following release (v0.4.0). I'll be reviewing this shortly.

I totally agree, this can be potentially disruptive, so I'd also prefer to be cautious and eventually rebase this onto #172

**Review by diegoferigo — CHANGES_REQUESTED** [review]
I've just finished a first review pass, things look generally good. I'll going to request some changes, and make a second pass later. I still need to weight some consequences especially on AD performance.

Here below my comments / questions:

- Most of the match-case seem to me 1:1 with the new lax.switch functions. Are there any sections in which you had to modify the original code? I checked the diff, but since it's quite large, I could have missed something. 
- In most cases, this PR does not alter significantly the readability of the code, I like that. We already had an extra indentation level due to the match-case statements, and the new functions use the same. There are cases, however, where the logic got much more complex and indented, like in `JaxSimModelReferences`. Let's keep this in mind in case we want to refactor it with a more simple approach.
- Do you think it could be helpful adding a new `jaxsim.typing.VelRepr` variable that points to `int`? Since this removed enum are part of the core public APIs, I'd prefer to make clear that those are not generic integer types.

File: src/jaxsim/api/references.py [hunk 1]
```diff
@@ -178,48 +180,80 @@ def link_forces(
...
+                        array=W_f_L,
+                        other_representation=velocity_representation,
+                        transform=W_H_L,
+                        is_force=True,
+                    )
+                )(W_f_L, W_H_L)
+
+            # The f_L output is either L_f_L or LW_f_L, depending on the representation.
+            W_H_L = js.model.forward_kinematics(
+                model=model, data=data or JaxSimModelData.zero(model=model)
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/references.py:244]: The logic should fail before, but since both branches are compiled, JAX was complaining that `data` was None

**Comment by flferretti** [comment]
> Most of the match-case seem to me 1:1 with the new lax.switch functions. Are there any sections in which you had to modify the original code? I checked the diff, but since it's quite large, I could have missed something.

No, they should be equivalent

> In most cases, this PR does not alter significantly the readability of the code, I like that. We already had an extra indentation level due to the match-case statements, and the new functions use the same. There are cases, however, where the logic got much more complex and indented, like in `JaxSimModelReferences`. Let's keep this in mind in case we want to refactor it with a more simple approach.

Yes, the logic got more complex since I needed to somehow check the values of some parameters. We can think of a smarter solution to handle that in the future.

> Do you think it could be helpful adding a new jaxsim.typing.VelRepr variable that points to int? Since this removed enum are part of the core public APIs, I'd prefer to make clear that those are not generic integer types.

Totally yes! I'll a commit for that

File: src/jaxsim/api/references.py [hunk 2]
```diff
@@ -366,22 +400,33 @@ def replace(forces: jtp.MatrixLike) -> JaxSimModelReferences:
...
+                true_fun=inertial,
+                false_fun=lambda: jax.pure_callback(
+                    callback=lambda: (_ for _ in ()).throw(
+                        ValueError(
+                            "Missing model to use a representation different from `VelRepr.Inertial`"
+                        )
+                    ),
+                    result_shape_dtypes=self,
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/references.py]: Done in https://github.com/ami-iit/jaxsim/pull/160/commits/432ec2c63f2f7d99d4d71d887f31c27a05689cbb

File: src/jaxsim/api/references.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/references.py]: Check https://github.com/ami-iit/jaxsim/pull/160/commits/432ec2c63f2f7d99d4d71d887f31c27a05689cbb

File: src/jaxsim/api/references.py [hunk 3]
```diff
@@ -178,48 +180,80 @@ def link_forces(
...
+                true_fun=inertial,
+                false_fun=lambda: jax.pure_callback(
+                    callback=lambda: (_ for _ in ()).throw(
+                        ValueError(
+                            "Missing model to use a representation different from `VelRepr.Inertial`"
+                        )
+                    ),
+                    result_shape_dtypes=self.input.physics_model.f_ext,
+                ),
+            )
```
[the remark this answers is not in the document]
Comment (flferretti) [reply src/jaxsim/api/references.py:201]: I'll leave this unresolved

[the remark this answers is not in the document]
Comment (diegoferigo) [reply src/jaxsim/api/references.py:201]: Got back on this in https://github.com/ami-iit/jaxsim/pull/181.

**Comment by flferretti** [comment]
I'll rebase this right before the next release @diegoferigo

**Comment by flferretti** [comment]
Checks are failing due to a timeout of the CI

**Comment by flferretti** [comment]
Closing without merging as this won't bring any significant performance/feature improvement
